crates/ide-completion/src/render/function.rs

```diff
@@ -153,6 +154,34 @@ fn render(
     item
 }
 
+/// When typing `::` of a type, the preferred order is:
+/// * Constructors: new like functions to be able to create the type,
+/// * Builder Methods,
+/// * Constructors that take args: Any other function that creates Self
+/// * Regular methods & Associated functions
+///
+fn calculate_bonus(ctx: &RenderContext<'_>, func: hir::Function, db: &dyn HirDatabase) -> u32 {
+    if ctx.token().kind() != syntax::SyntaxKind::COLON2 || func.self_param(db).is_some() {
+        return 0;
+    }
+
+    let mut bonus = 0;
+
+    let has_args = !func.assoc_fn_params(db).is_empty();
+    let ret_type = func.ret_type(db);
+    if !has_args && !ret_type.is_unit() {
+        // fn() -> A
+        bonus += 30;
+    } else if has_args && !ret_type.is_unit() {
+        // fn(..) -> A
+        bonus += 20;
```

Comment:
These need to check that the return type is the same as the `Self` type of the trait or impl the function is a part of, otherwise these are generally associated functions being preferred instead of just constructors.